libraries/libstratosphere/source/fssystem/fssystem_local_file_system.cpp

```diff
@@ -0,0 +1,1751 @@
+        R_TRY(this->ResolveFullPath(std::addressof(native_path), path, MaxFilePathLength, 0, true));
+
+        /* Delete the directory, retrying on target locked. */
+        auto delete_impl = [&] () -> Result {
+            #if defined(ATMOSPHERE_OS_WINDOWS)
+            R_UNLESS(::RemoveDirectoryW(native_path.get()), ConvertLastErrorToResult());
+            #else
+            R_UNLESS(::rmdir(native_path.get()) == 0, ConvertErrnoToResult(ErrnoSource_Rmdir));
+            #endif
+
+            R_SUCCEED();
+        };
+
+        /* Perform the delete. */
+        R_TRY(fssystem::RetryToAvoidTargetLocked(delete_impl));
+
+        /* Wait for the deletion to complete. */
+        R_RETURN(WaitDeletionCompletion(native_path.get()));
+    }
+
+    Result LocalFileSystem::DoDeleteDirectoryRecursively(const fs::Path &path) {
+        /* Resolve the path. */
+        NativePathBuffer native_path;
+        R_TRY(this->ResolveFullPath(std::addressof(native_path), path, MaxFilePathLength, 0, true));
+
+        /* Delete the directory. */
+        R_RETURN(this->DeleteDirectoryRecursivelyInternal(native_path.get(), true));
+    }
+
+    Result LocalFileSystem::DoRenameFile(const fs::Path &old_path, const fs::Path &new_path) {
+        /* Resolve the old path. */
+        NativePathBuffer native_old_path;
+        R_TRY(this->ResolveFullPath(std::addressof(native_old_path), old_path, MaxFilePathLength, 0, true));
+
+        /* Resolve the new path. */
+        NativePathBuffer native_new_path;
+        R_TRY(this->ResolveFullPath(std::addressof(native_new_path), new_path, MaxFilePathLength, 0, false));
+
+        /* Check that the old path is a file. */
+        fs::DirectoryEntryType type;
+        R_TRY(GetEntryTypeImpl(std::addressof(type), native_old_path.get()));
+        R_UNLESS(type == fs::DirectoryEntryType_File, fs::ResultPathNotFound());
+
+        /* Perform the rename. */
+        auto rename_impl = [&]() -> Result {
+            #if defined(ATMOSPHERE_OS_WINDOWS)
+            if (!::MoveFileW(native_old_path.get(), native_new_path.get())) {
+                R_TRY_CATCH(ConvertLastErrorToResult()) {
+                    R_CATCH(fs::ResultTargetLocked) {
+                        /* If we're performing a self rename, succeed. */
+                        R_SUCCEED_IF(::wcscmp(native_old_path.get(), native_new_path.get()) == 0);
+
+                        /* Otherwise, check if the new path already exists. */
+                        const auto attr = ::GetFileAttributesW(native_new_path.get());
+                        R_UNLESS(attr == INVALID_FILE_ATTRIBUTES, fs::ResultPathAlreadyExists());
+
+                        /* Return the original result. */
+                        R_THROW(R_CURRENT_RESULT);
+                    }
+                } R_END_TRY_CATCH;
+            }
+            #else
+            {
+                /* ::rename() will destroy an existing file at new path...so check for that case ahead of time. */
```

Comment:
`renameat2` has a flag to error out if the destination exists (`RENAME_NOREPLACE`)

Reply:
I could add this, but note that I would have to ifdef because macOS does not have `renameat2`.